File: cpp/HashTable/simple_hash.cpp
/* Additive hash */
int addHash(string key) {
  long long hash = 0;
  const int MODULUS = 1000000007;
  for (unsigned char c : key) {
    hash = (hash + (int)c) % MODULUS;
  }
  return (int)hash;
}

/* Multiplicative hash */
int mulHash(string key) {
  long long hash = 0;
  const int MODULUS = 1000000007;
  for (unsigned char c : key) {
    hash = (31 * hash + (int)c) % MODULUS;
  }
  return (int)hash;
}

/* XOR hash */
int xorHash(string key) {
  int hash = 0;
  const int MODULUS = 1000000007;
  for (unsigned char c : key) {
    hash ^= (int)c;
  }
  return hash & MODULUS;
}

/* Rotational hash */
int rotHash(string key) {
  long long hash = 0;
  const int MODULUS = 1000000007;
  for (unsigned char c : key) {
    hash = ((hash << 4) ^ (hash >> 28) ^ (int)c) % MODULUS;
  }
  return (int)hash;
}

File: java/hash_table/simple_hash.java
/* Additive hash */
int addHash(String key) {
    long hash = 0;
    final int MODULUS = 1000000007;
    for (char c : key.toCharArray()) {
        hash = (hash + (int) c) % MODULUS;
    }
    return (int) hash;
}

/* Multiplicative hash */
int mulHash(String key) {
    long hash = 0;
    final int MODULUS = 1000000007;
    for (char c : key.toCharArray()) {
        hash = (31 * hash + (int) c) % MODULUS;
    }
    return (int) hash;
}

/* XOR hash */
int xorHash(String key) {
    int hash = 0;
    final int MODULUS = 1000000007;
    for (char c : key.toCharArray()) {
        hash ^= (int) c;
    }
    return hash & MODULUS;
}

/* Rotational hash */
int rotHash(String key) {
    long hash = 0;
    final int MODULUS = 1000000007;
    for (char c : key.toCharArray()) {
        hash = ((hash << 4) ^ (hash >> 28) ^ (int) c) % MODULUS;
    }
    return (int) hash;
}
